226: support locale variants

toolstack opened the pull request:
Resolves #226 .

toolstack committed:
First cut at supporting locale variants.

toolstack committed:
Add root translation highlighting in translations list.

Also add a link back to the root translation in the detail view.

toolstack committed:
Updated gp_for_translation_clauses filter to match new root query format.

toolstack committed:
Update more root/variant locales.

swissspidy commented on locales/locales.php:
```diff
@@ -972,6 +1010,8 @@ public function __construct() {
 		$gsw->country_code = 'ch';
 		$gsw->wp_locale = 'gsw';
 		$gsw->slug = 'gsw';
+		$gsw->variant_root = $de->slug;
```
Please note that while German (Switzerland) as a variant of standard German makes sense, Swiss German (gsw) is something completely different. It's a dialect (well, more than 26 different dialects to be precise) that has nothing in common with regular German. As such, `$gsw` shouldn't be a variant of `$de`.

pixelverbieger replied:
… while _de_DE_formal_ is missing and should be added to the list of variants for German

swissspidy replied:
That would be handled somewhere else though 🙂 `de_DE_formal` is not listed in `locales.php` after all, it's probably just a WordPress.org thing.

toolstack replied:
Thanks @swissspidy, I'll remove the relationship.

@pixelverbieger I'm sure there will be many new variants added, like de_DE_formal once the the PR is merged, currently swsisspidy is correct that the formal variants are handled by translate.w.org separately.

toolstack committed:
Change color of root translations and add filter for locales.

toolstack committed:
Fix filter function call.

toolstack committed:
Make sure apply_filters() exists before calling it.

If someone is using the locale file outside of WordPress, the lack of apply_filters() would cause a fatal error.

toolstack committed:
Fix function name in function_exists() call.

And coding standard updates.

toolstack committed:
Update Chinese root/variants.

And coding standard updates.

toolstack committed:
Fix test case.

ocean90 commented on gp-templates/translation-row.php:
```diff
@@ -178,6 +178,32 @@
 
+			<?php if ( property_exists( $t, 'root_translation_set_id' ) ) : ?>
+			<dl>
+				<dt><?php _e( 'Root Translaton:', 'glotpress' ); ?></dt>
```
Typo

toolstack replied:
Fixed.

toolstack committed:
Update phpcs rules.

toolstack committed:
Coding standard and merge conflict updates.

HughP commented:
I’m not sure why the code invokes iso 629-2 codes.  BCP 47 Tags should be used. 

I don’t have a problem with the fallbacks chosen. They make sense. But it seems that the coding seems more complex than needed. ISO 639-2 and 639-3 are alligned for the purposes of locale codes CLDR or Iana should have the already Approved codes.

xavivars commented:
Any idea of progress on this PR? 

While discussing about the details could take forever (this language is/is-not a variant of another language), the functionality itself it's extremely useful for cases that are clear. For example, we now have *Valencian (ca-valencia, Catalan as spoken in Valencia), and falling back to English is simply not an option: there aren't enough translators to keep all themes/plugins up to date in both variants.

toolstack commented:
It will be merged as part of 3.0, which is scheduled for a Dec 20th release.

toolstack committed:
Added rejected to the legend.

xavivars commented:
I guess ca-valencia will need to be added as a variant of ca in locales.php for this case to work, right?

toolstack commented:
@xavivars Correct, it can be done for a site by either hooking in to the `gp_locale_definitions_arrary` filter or creating a custom locales file.

toolstack committed:
Add test case.

toolstack committed:
Coding standard updates.

toolstack committed:
Add a no variants version of the locale file.

This can be used with the GP_LOCALES_PATH define to  effectively disable variants.

toolstack committed:
Fixed typos in comments.
